src/sentry/users/models/identity.py

```diff
@@ -0,0 +1,216 @@
+from __future__ import annotations
+
+import logging
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Any, ClassVar
+
+from django.conf import settings
+from django.db import IntegrityError, models
+from django.db.models import Q, QuerySet
+from django.utils import timezone
+
+from sentry import analytics
+from sentry.backup.scopes import RelocationScope
+from sentry.db.models import (
+    ArrayField,
+    BoundedPositiveIntegerField,
+    FlexibleForeignKey,
+    Model,
+    control_silo_model,
+)
+from sentry.db.models.fields.jsonfield import JSONField
+from sentry.db.models.manager.base import BaseManager
+from sentry.integrations.types import ExternalProviders
+from sentry.users.services.user import RpcUser
+
+if TYPE_CHECKING:
+    from sentry.identity.base import Provider
+    from sentry.identity.services.identity import RpcIdentityProvider
+    from sentry.users.models.user import User
+
+logger = logging.getLogger(__name__)
+
+
+# TODO(dcramer): pull in enum library
+class IdentityStatus:
+    UNKNOWN = 0
+    VALID = 1
+    INVALID = 2
+
+
+@control_silo_model
+class IdentityProvider(Model):
+    """
+    An IdentityProvider is an instance of a provider.
+
+    The IdentityProvider is unique on the type of provider (eg github, slack,
+    google, etc).
+
+    A SAML identity provide might look like this, type: onelogin, instance:
+    acme-org.onelogin.com.
+    """
+
+    __relocation_scope__ = RelocationScope.Excluded
+
+    type = models.CharField(max_length=64)
+    config: models.Field[dict[str, Any], dict[str, Any]] = JSONField()
+    date_added = models.DateTimeField(default=timezone.now, null=True)
+    external_id = models.CharField(max_length=64, null=True)
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_identityprovider"
+        unique_together = (("type", "external_id"),)
+
+    def get_provider(self) -> Identity:
+        from sentry.identity import get
+
+        return get(self.type)
```

Comment:
I interpreted the `get()` function returning a `cls("Identity")` which would just be an Identity class... (???)

Reply:
Wouldn't this be `IdentityProvider`. The `get()` will return an instance of the registered types (Providers) based on the string name provided.

Providers are registered here https://github.com/getsentry/sentry/blob/4fb67fb784549608feffa3798db2dadd2a4fb6a2/src/sentry/identity/__init__.py#L25-L34

The `Slack` provider is found https://github.com/getsentry/sentry/blob/6ef91129b58b3627e9705d0ad71c1bf458567919/src/sentry/identity/slack/provider.py#L5

and the class tree for those ends in `Provider` https://github.com/getsentry/sentry/blob/6ef91129b58b3627e9705d0ad71c1bf458567919/src/sentry/identity/base.py#L7

Reply:
Ohhhh, gotchas. I completely missed the whole provider registration and that the `self.values` in `IdentityManager` was populated with Providers from the registration.